pkg/vectorize/ceil/ceil.go

```diff
@@ -0,0 +1,381 @@
+func ceilUint16Pure(xs, rs []uint16, digits int64) []uint16 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxUint16digits:
+		scale := uint16(ScaleTable[-digits])
+		for i := range xs {
+			t := xs[i] % scale
+			s := xs[i]
+			if t != 0 {
+				s -= t
+				rs[i] = (s + scale) / scale * scale
+			} else {
+				rs[i] = xs[i]
+			}
+		}
+	case digits <= -MaxInt8digits:
+		for i := range xs {
+			rs[i] = 0
+		}
+	}
+	return rs
+}
+
+func CeilUint32(xs, rs []uint32, digits int64) []uint32 {
+	return ceilUint32(xs, rs, digits)
+}
+
+func ceilUint32Pure(xs, rs []uint32, digits int64) []uint32 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxInt16digits:
+		scale := uint32(ScaleTable[-digits])
+		for i := range xs {
+			t := xs[i] % scale
+			s := xs[i]
+			if t != 0 {
+				s -= t
+				rs[i] = (s + scale) / scale * scale
+			} else {
+				rs[i] = xs[i]
+			}
+		}
+	case digits <= -MaxInt8digits:
+		for i := range xs {
+			rs[i] = 0
+		}
+	}
+	return rs
+}
+
+func CeilUint64(xs, rs []uint64, digits int64) []uint64 {
+	return ceilUint64(xs, rs, digits)
+}
+
+func ceilUint64Pure(xs, rs []uint64, digits int64) []uint64 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxUint16digits:
+		scale := uint64(ScaleTable[-digits])
+		for i := range xs {
+			t := xs[i] % scale
+			s := xs[i]
+			if t != 0 {
+				s -= t
+				rs[i] = (s + scale) / scale * scale
+			} else {
+				rs[i] = xs[i]
+			}
+		}
+	case digits <= -MaxUint8digits:
+		for i := range xs {
+			rs[i] = 0
+		}
+	}
+	return rs
+}
+
+func CeilInt8(xs, rs []int8, digits int64) []int8 {
+	return ceilInt8(xs, rs, digits)
+}
+
+func ceilInt8Pure(xs, rs []int8, digits int64) []int8 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits == -1 || digits == -2:
+		scale := int8(ScaleTable[-digits])
+		for i := range xs {
+			t := xs[i] % scale
+			s := xs[i]
+			if t != 0 {
+				s -= t
+				if s >= 0 && xs[i] > 0 {
+					rs[i] = (s + scale) / scale * scale
+				} else {
+					rs[i] = s
+				}
+			} else {
+				rs[i] = xs[i]
+			}
+		}
+	case digits <= -MaxUint8digits:
+		for i := range xs {
+			rs[i] = 0
+		}
+	}
+	return rs
+}
+
+func CeilInt16(xs, rs []int16, digits int64) []int16 {
+	return ceilInt16(xs, rs, digits)
+}
+
+func ceilInt16Pure(xs, rs []int16, digits int64) []int16 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxUint16digits:
+		scale := int16(ScaleTable[-digits])
+		for i := range xs {
+			t := xs[i] % scale
+			s := xs[i]
+			if t != 0 {
+				s -= t
+				if s >= 0 && xs[i] > 0 {
+					rs[i] = (s + scale) / scale * scale
+				} else {
+					rs[i] = s
+				}
+			} else {
+				rs[i] = xs[i]
+			}
+		}
+	case digits <= -MaxUint8digits:
+		for i := range xs {
+			rs[i] = 0
+		}
+	}
+	return rs
+}
+
+func CeilInt32(xs, rs []int32, digits int64) []int32 {
+	return ceilInt32(xs, rs, digits)
+}
+
+func ceilInt32Pure(xs, rs []int32, digits int64) []int32 {
+	switch {
+	case digits >= 0:
+		return xs
+	case digits > -MaxUint16digits:
```

Comment:
perhaps this is digits > -MaxInt32digits

Reply:
problem solved